### PR TITLE
Add Nonnull and Nullable annotations and minor behaviour fixes

### PR DESCRIPTION
**Pull request by mmedenjak**
Adds Nonnull and Nullable annotations to HazelcastInstance,
CardinalityEstimator, IExecutor, DurableExecutor, QuorumService and some
others.
Aligned some behaviour of client and member-side when it comes to
behaviour with null parameters.
Added FunctionalInterface to MemberSelector and cleaned up some JDK8
code.

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3115

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/quorum/impl/QuorumServiceImpl.java [hunk 1]
```diff
@@ -224,12 +226,18 @@ public void ensureQuorumPresent(Operation op) {
...
+    @Override
+    public void ensureQuorumPresent(@Nullable String quorumName,
+                                    @Nonnull QuorumType requiredQuorumPermissionType) {
+        checkNotNull(requiredQuorumPermissionType, "requiredQuorumPermissionType cannot be null!");
         if (isInactive() || quorumName == null) {
             return;
         }
 
         QuorumImpl definedQuorum = quorums.get(quorumName);
+        if (definedQuorum == null) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/quorum/impl/QuorumServiceImpl.java]: This was a fix where there was a NPE if there was no configured quorum with the provided `quorumName`. I assumed the same behaviour as when there is no configured qourum (`isInactive() == true` above).

File: hazelcast/src/main/java/com/hazelcast/cluster/MemberSelector.java [hunk 2]
```diff
@@ -31,6 +31,7 @@
  *     }
  * }</pre>
  */
+@FunctionalInterface
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/cluster/MemberSelector.java]: Added functional interface

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientExecutorServiceProxy.java [hunk 3]
```diff
@@ -573,7 +643,9 @@ public void onFailure(Throwable t) {
         private final Map<Member, Object> values;
         private final AtomicInteger members;
 
-        private MultiExecutionCallbackWrapper(int memberSize, MultiExecutionCallback multiExecutionCallback) {
+        private MultiExecutionCallbackWrapper(int memberSize,
+                                              @Nonnull MultiExecutionCallback multiExecutionCallback) {
+            checkNotNull(multiExecutionCallback, "multiExecutionCallback must not be null");
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientExecutorServiceProxy.java]: Aligning behaviour with member-side. NPE should be thrown as soon as the input is not valid and not later in a separate thread.

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientExecutorServiceProxy.java [hunk 4]
```diff
@@ -402,20 +455,26 @@ private Object getTaskPartitionKey(Object task) {
...
         String uuid = getUUID();
         int partitionId = getPartitionId(key);
         ClientMessage request = ExecutorServiceSubmitToPartitionCodec.encodeRequest(name, uuid, task, partitionId);
         ClientInvocationFuture f = invokeOnPartitionOwner(request, partitionId);
 
-        ClientDelegatingFuture<T> delegatingFuture = new ClientDelegatingFuture<T>(f, getSerializationService(),
+        ClientDelegatingFuture<T> delegatingFuture = new ClientDelegatingFuture<>(f, getSerializationService(),
                 message -> ExecutorServiceSubmitToPartitionCodec.decodeResponse(message).response);
-        delegatingFuture.andThen(callback);
+        if (callback != null) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientExecutorServiceProxy.java]: This fix and similar ones below are aligning behaviour with member-side where the callback is ignored if it is `null`.

File: hazelcast/src/main/java/com/hazelcast/quorum/impl/QuorumServiceImpl.java (hunk 1 above)
Comment (vojtechtoman) [reply hazelcast/src/main/java/com/hazelcast/quorum/impl/QuorumServiceImpl.java]: Just a note for posterity: in `SplitBrainProtectionServiceImpl` now.

**Comment by asimarslan** [comment]
We need to apply the same idea to generated codecs.

**Comment by mmedenjak** [comment]
Thank you for the reviews @vojtechtoman @asimarslan !

**Pull request merged**